omniledger/contracts/coins.go

```diff
@@ -58,23 +59,28 @@ func (s safeUint64) sub(a uint64) (safeUint64, error) {
 // You can only delete a contractCoin instance if the account is empty.
 func ContractCoin(cdb omniledger.CollectionView, inst omniledger.Instruction, c []omniledger.Coin) (sc []omniledger.StateChange, cOut []omniledger.Coin, err error) {
 	cOut = c
+
+	var darcID darc.ID
+	_, _, darcID, err = cdb.GetValues(inst.InstanceID.Slice())
+	if err != nil {
+		return
+	}
+
 	switch inst.GetType() {
 	case omniledger.SpawnType:
-		// Spawn creates a new coin account as a separate instance. The subID is
-		// taken from the hash of the instruction.
-		ca := omniledger.InstanceID{
-			DarcID: inst.InstanceID.DarcID,
-			SubID:  omniledger.NewSubID(inst.Hash()),
-		}
-		log.Lvlf3("Spawing coin to %x", ca.Slice())
+		// Spawn creates a new coin account as a separate instance.
+		ca := omniledger.InstanceIDFromSlice(inst.Hash())
```

Comment:
Why is it `inst.Hash()` instead of `inst.DeriveID(ContractCoinID)`?

Reply:
Yeah, exactly. I was struggling with this too.

When I first realised we needed DeriveID it was to solve the case where you need to create a new object in the collection in order to help store/index the current item. I.e. while writing a new event, you need a new time bucket to store it in.

But you are proposing that new instance ID's should *always* come from DeriveID, which might be simpler to explain to people.

I do think it is regrettable that client side code which wants to predict what key a given object will be stored at needs to know the magic string to give to it's own invocation of DeriveId. To me that's kind of the internals of the contract and none of the client's business.

But in eventlog, we already violate that, because the instance ID of a newly spawned eventlog is not inst.Hash(), but is in fact inst.DeriveId("eventlog").

Reply:
If we say that `DeriveID` for a new instance should take the `ContractID` of that the new instance?